tensorboard/plugins/metrics/metrics_plugin.py

```diff
@@ -0,0 +1,590 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""The TensorBoard metrics plugin."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import imghdr
+import json
+
+from werkzeug import wrappers
+
+from tensorboard import errors
+from tensorboard import plugin_util
+from tensorboard.backend import http_util
+from tensorboard.data import provider
+from tensorboard.plugins import base_plugin
+from tensorboard.plugins.histogram import metadata as histogram_metadata
+from tensorboard.plugins.image import metadata as image_metadata
+from tensorboard.plugins.metrics import metadata
+from tensorboard.plugins.scalar import metadata as scalar_metadata
+
+
+_IMGHDR_TO_MIMETYPE = {
+    "bmp": "image/bmp",
+    "gif": "image/gif",
+    "jpeg": "image/jpeg",
+    "png": "image/png",
+    "svg": "image/svg+xml",
+}
+
+_DEFAULT_IMAGE_MIMETYPE = "application/octet-stream"
+
+_SINGLE_RUN_PLUGINS = frozenset(
+    [histogram_metadata.PLUGIN_NAME, image_metadata.PLUGIN_NAME]
+)
+
+_SAMPLED_PLUGINS = frozenset([image_metadata.PLUGIN_NAME])
+
+
+def _get_tag_description_info(mapping):
+    """Gets maps from tags to descriptions, and descriptions to runs.
+
+    Args:
+        mapping: a nested map `d` such that `d[run][tag]` is a time series
+          produced by DataProvider's `list_*` methods.
+
+    Returns:
+        A tuple containing
+            tag_to_descriptions: A map from tag strings to a set of description
+                strings.
+            description_to_runs: A map from description strings to a set of run
+                strings.
+    """
+    tag_to_descriptions = collections.defaultdict(set)
+    description_to_runs = collections.defaultdict(set)
+    for (run, tag_to_content) in mapping.items():
+        for (tag, metadatum) in tag_to_content.items():
+            description = metadatum.description
+            if len(description):
+                tag_to_descriptions[tag].add(description)
+                description_to_runs[description].add(run)
+
+    return tag_to_descriptions, description_to_runs
+
+
+def _build_combined_description(descriptions, description_to_runs):
+    """Creates a single description from a set of descriptions.
+
+  Descriptions may be composites when a single tag has different descriptions
```

Comment:
All the docstring formatting is borked. Could you please run
`docformatter` once? I did this when first adding Black in #2967 and it
worked like a charm.

Reply:
Done.  It did work like a charm!